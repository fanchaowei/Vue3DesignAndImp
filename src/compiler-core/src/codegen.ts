// 将 JavaScript AST 转换为渲染函数的代码
export function generate(node: any) {
  // 全局的上下文对象
  const context = {
    code: '',
    // 在生成代码时，通过调用 push 函数完成代码的拼接
    push(code: string) {
      context.code += code
    },
    // 当前缩进的级别，一个级别为两个空格
    currentIndent: 0,
    // 用于换行的函数
    // 换行后，要保持缩进
    newline() {
      context.code += '\n' + '  '.repeat(context.currentIndent)
    },
    // 用来缩进，即让 currentIndent 自增后，调用换行函数
    indent() {
      context.currentIndent++
      context.newline()
    },
    // 取消缩进，让 currentIdent 自减后，调用换行函数
    deIndent() {
      context.currentIndent--
      context.newline()
    }
  }

  // 调用 genNode 函数完成代码生成的工作
  genNode(node, context)

  return context.code
}

// 处理 JavaScript AST
function genNode(node: any, context: any) {
  // 按照 node.type 调用对应的转换方法
  switch (node.type) {
    case 'FunctionDecl':
      genFunctionDecl(node, context)
      break
    case 'ReturnStatement':
      genReturnStatement(node, context)
      break
    case 'CallExpression':
      genCallExpression(node, context)
      break
    case 'StringLiteral':
      genStringLiteral(node, context)
      break
    case 'ArrayExpression':
      genArrayExpression(node, context)
      break
  }
}
// 处理数组，生成对应的 JavaScript 代码
function genArrayExpression(node: any, context: any) {
  const { push } = context
  push(`[`)
  genNodeList(node.elements, context)
  push(`]`)
}
// 处理文本，生成对应的 JavaScript 代码
function genStringLiteral(node: any, context: any) {
  const { push } = context
  push(`'${node.value}'`)
}
// 处理函数调用，生成对应的 JavaScript 代码
function genCallExpression(node: any, context: any) {
  const { push } = context
  const { callee, arguments: args } = node
  push(`${callee.name}(`)
  // 生成参数
  genNodeList(args, context)
  push(`)`)
}
// 处理函数返回，生成对应的 JavaScript 代码
function genReturnStatement(node: any, context: any) {
  const { push } = context
  // 追加 return 关键字和空格
  push(`return `)
  genNode(node.return, context)
}
// 处理函数的声明，生成对应的 JavaScript 代码
function genFunctionDecl(node: any, context: any) {
  const { push, indent, deIndent } = context
  push(`function ${node.id.name}`)
  push(`(`)
  // 生成函数的参数
  genNodeList(node.params, context)
  push(`) `)
  push(`{`)
  // 缩进
  indent()
  // 依次将内部的内容传入到 genNode 中生成对应的渲染函数代码
  node.body.forEach((n: any) => genNode(n, context))
  deIndent()
  push(`}`)
}

// 生成函数声明的参数或数组的内容
function genNodeList(nodes: any, context: any) {
  const { push } = context
  // 遍历参数
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i]
    // 将每个参数节点都传入 genNode 中进行处理
    genNode(node, context)
    if (i < nodes.length - 1) {
      push(', ')
    }
  }
}
