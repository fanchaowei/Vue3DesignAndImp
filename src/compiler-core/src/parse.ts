// 定义状态机的状态
const State = {
  initial: 1, // 初始状态
  tagOpen: 2, // 标签开始状态
  tagName: 3, // 标签名称状态
  text: 4, // 文本状态
  tagEnd: 5, // 标签结束状态
  tagEndName: 6 // 结束标签名称状态
}

// 辅助函数，用于判断是否是字母
function isAlpha(char: String) {
  return (char >= 'a' && char <= 'z') || (char >= 'A' && char <= 'Z')
}

// 接受模板字符串作为参数，并将模板切割为 Token 返回
export function tokenize(str: String) {
  // 状态机的当前状态，默认在初始状态
  let currentState = State.initial
  // 用于缓存字符
  const chars = []
  // 用于存放生成的 token
  const tokens = []

  // 开启自动机循环
  while (str) {
    // 查看当前 str 的第一个字母
    const char = str[0]
    // 根据当前的状态，来处理 char
    switch (currentState) {
      // 当前为初始状态
      case State.initial:
        if (char === '<') {
          // 如果是 < ，则切换到标签开始状态，并排出该字符
          currentState = State.tagOpen
          str = str.slice(1)
        } else if (isAlpha(char)) {
          // 如果是字母，则切换到文本状态
          currentState = State.text
          // 将当前字母添加到缓存数组
          chars.push(char)
          // 排出该字符
          str = str.slice(1)
        }
        break
      // 当前为标签开始状态
      case State.tagOpen:
        if (isAlpha(char)) {
          // 如果是字母，切换到标签名称状态
          currentState = State.tagName
          // 存入缓存数组
          chars.push(char)
          str = str.slice(1)
        } else if (char === '/') {
          // 如果是 / ，进入结束标签状态
          currentState = State.tagEnd
          str.slice(1)
        }
        break
      // 标签名称状态
      case State.tagName:
        if (isAlpha(char)) {
          // 如果是字母，存入缓存数组
          chars.push(char)
          str = str.slice(1)
        } else if (char === '>') {
          // 如果是 > ，切换到初始状态
          currentState = State.initial
          // 创建一个标签 token，讲当前缓存的信息存入。
          tokens.push({
            type: 'tag',
            name: chars.join('')
          })
          // 字符缓存清空
          chars.length = 0
          str.slice(1)
        }
        break
      // 文本状态
      case State.text:
        if (isAlpha(char)) {
          // 如果是字母，存入缓存数组
          chars.push(char)
          str = str.slice(1)
        } else if (char === '<') {
          // 如果是 < ，说明已经结束了文本状态，进入结束标签状态
          currentState = State.text
          // 创建一个文本 token，讲当前缓存信息存入
          tokens.push({
            type: 'text',
            content: chars.join('')
          })
          // 清空缓存
          chars.length = 0
          str = str.slice(1)
        }
        break
      // 标签结束状态
      case State.tagEnd:
        if (isAlpha(char)) {
          // 如果是字母，进入结束标签名称状态
          currentState = State.tagEndName
          chars.push(char)
          str = str.slice(1)
        }
        break
      // 结束标签名称状态
      case State.tagEndName:
        if (isAlpha(char)) {
          // 如果是字母，存入缓存数组
          chars.push(char)
          str = str.slice(1)
        } else if (char === '>') {
          // 如果是 > ，回到初始状态
          currentState = State.initial
          // 创建一个结束标签名称 token
          tokens.push({
            type: 'tagEnd',
            name: chars.join('')
          })
          chars.length = 0
          str = str.slice(1)
        }
        break
    }
  }
  return tokens
}
