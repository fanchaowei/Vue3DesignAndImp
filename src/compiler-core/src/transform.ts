import {
  createArrayExpression,
  createCallExpression,
  createStringLiteral,
  createIdentifier,
  dump
} from './utils'

// 用于转换 AST
export function traverseNode(ast: any, context: any) {
  // 存储当前转换中的节点
  context.currentNode = ast
  // 增加退出阶段的回调函数数组
  const exitFns = []
  // 从 context 中获取 ast 的处理数组
  const transforms = context.nodeTransforms
  for (let i = 0; i < transforms.length; i++) {
    // 循环转换数组，将当前的节点传入进行处理
    const onExit = transforms[i](context.currentNode, context)
    // 转换函数可以返回一个函数，作为结束阶段的回调函数
    if (onExit) {
      exitFns.push(onExit)
    }
    // 因为每个转换处理函数都可能移除节点，所以每次执行转换函数后，都得判断一下是否还存在当前节点
    if (!context.currentNode) return
  }

  // 有 children 则递归
  const children = context.currentNode.children
  if (children) {
    children.forEach((c: any, index: number) => {
      // 存储父级节点
      context.parent = context.currentNode
      // 存储索引
      context.childIndex = index
      traverseNode(c, context)
    })
  }

  // 反序执行当前节点存储的回调函数
  let i = exitFns.length
  while (i--) {
    exitFns[i]()
  }
}

export function transforms(ast: any) {
  // 创建 context 对象
  const context: any = {
    // 用于存储当前正在转换的节点
    currentNode: null,
    // 用于存储当前节点在父节点的 children 中的位置索引
    childIndex: 0,
    // 用于存储当前转换节点的父节点
    parent: null,
    // 处理函数数组，提供给 traverseNode 调用
    nodeTransforms: [transformRoot, transformElement, transformText],
    // 替换节点
    replaceNode(node: any) {
      //替换节点，我们需要通过父节点的 children 找到当前节点来替换。
      context.parent.children[context.childIndex] = node
      // currentNode 要修改为替换的节点
      context.currentNode = node
    },
    // 用于删除当前节点
    removeNode() {
      // 调用 splice  方法删除当前节点
      context.parent.children.splice(context.childIndex, 1)
      // 将 context.currentNode 置空
      context.currentNode = null
    }
  }
  // 调用 traverseNode 进行转换
  traverseNode(ast, context)
  console.log(dump(ast))
}

// 转换元素节点
function transformElement(node: any) {
  // 将转换代码编写再退出阶段的回调函数中
  // 这样可以保证该标签的所有子节点全部被处理完毕
  return () => {
    // 如果被转换的节点不是元素节点，则什么都不做
    if (node.type !== 'Element') {
      return
    }

    // 创建该节点的 h 函数的描述
    const callExp = createCallExpression(createIdentifier('h'), [
      createStringLiteral(node.tag)
    ])

    node.children.length === 1
      ? // 如果当前标签节点只有一个子节点，则直接使用子节点的 jsNode 作为参数
        callExp.arguments.push(node.children[0].jsNode)
      : // 如果有多个节点，则创建数组类型的 js AST 节点
        callExp.arguments.push(
          // 数组的每个元素都是子节点的 jsNode
          createArrayExpression(node.children.map((c: any) => c.jsNode))
        )
    // 将该标签的 JavaScript AST 存入 jsNode
    node.jsNode = callExp
  }
}
// 转换文本节点
function transformText(node: any, context: any) {
  if (node.type === 'Text') {
    // 如果是文本，则创建一个文本类型的 JavaScript AST 节点添加到 node.jsNode 属性下
    node.jsNode = createStringLiteral(node.content)
  }
}
// 转换 Root 根节点
function transformRoot(node: any) {
  return () => {
    // 如果不是根节点，则什么都不做
    if (node.type !== 'Root') {
      return
    }

    // 获取 Root 节点下，实际的根节点。及 Root 的子节点
    // 这里暂不考虑有多个实际根节点的情况
    const vnodeJSAST = node.children[0].jsNode
    // 创建 render 函数的声明语句节点，将 vnodeJSAST 作为 render 函数体的返回语句
    node.jsNode = {
      type: 'FunctionDecl',
      id: createIdentifier('render'),
      params: [],
      body: [
        {
          type: 'ReturnStatement',
          return: vnodeJSAST
        }
      ]
    }
  }
}
