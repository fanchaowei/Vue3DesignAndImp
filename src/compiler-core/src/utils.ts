// 工具函数，用于打印 AST 节点的信息
export function dump(node: any, indent = 0) {
  // 节点的类型
  const type = node.type
  // 节点的描述，如果是 Root 节点则描述为空
  const desc =
    node.type === 'Root'
      ? ''
      : node.type === 'Element'
      ? node.tag
      : node.content
  console.log(`${'-'.repeat(indent)}${type}: ${desc}`)
  // 存在 children 则递归
  if (node.children) {
    node.children.forEach((n: any) => dump(n, indent + 2))
  }
}
