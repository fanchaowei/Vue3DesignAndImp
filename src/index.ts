export * from './reactivity'
export * from './runtime-dom'
export * from './temporary'
