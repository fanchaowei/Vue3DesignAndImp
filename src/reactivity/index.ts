export * from './reactive'
export * from './ref'
export { effect } from './effect'
export { computed } from './computed'
