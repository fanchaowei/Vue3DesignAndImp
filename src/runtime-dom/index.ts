import { createRenderer, Fragment } from '../runtime-core'

// 创建 DOM 元素
function createElement(tag: any) {
  return document.createElement(tag)
}
// 输入文本
function setElementText(el: any, text: any) {
  el.textContent = text
}
// 添加元素到父级元素
function insert(el: any, parent: any, auchor = null) {
  // 将 el 添加到 anthor 之前，倘若为 null ，则默认添加到最后，和 append() 一样。
  // 倘若给定的子节点是当前已存在的节点，则会将这个节点移动到锚点之前
  parent.insertBefore(el, auchor)
}
// 卸载元素
function unmount(vnode: any) {
  if (vnode.type === Fragment) {
    // Fragment 类型需要先依次卸载每个子节点
    vnode.children.forEach((c: any) => {
      unmount(c)
    })
    return
  } else if(typeof vnode.type === 'object') {
    // 对于组件，则是卸载组件所渲染的内容，即 subTree
    unmount(vnode.component.subTree)
    return
  }
  // 获取元素的父级
  const parent = vnode.el.parentNode
  if (parent) {
    // 卸载元素
    parent.removeChild(vnode.el)
  }
}
// 创建文本节点
function createText(text: string) {
  return document.createTextNode(text)
}
// 设置文本节点的内容
function setText(el: any, text: string) {
  el.nodeValue = text
}

// 处理元素 Attributes
function patchProps(el: any, key: any, prevValue: any, nextValue: any) {
  if (/^on/.test(key)) {
    // 判断是否是事件
    // 将 key 进行转化，如：onClick -> click
    const name = key.slice(2).toLowerCase()
    // 获取伪造事件集
    let invokers = el._vei || (el._vei = {})
    // 通过名称获取对应的伪造事件处理函数 invoker
    let invoker = invokers[name]
    // 判断是否存在新的事件
    if (nextValue) {
      if (!invoker) {
        // 如果不存在 invoker 则创建一个
        invoker = el._vei = (e: any) => {
          // e.timeStamp 是事件触发的时间
          // 如果事件发生的事件遭遇事件处理函数绑的时间，则不执行事件处理函数
          if (e.timeStamp < invoker.attached) return
          // invoker 内实际是调用 invoker.value
          // 而我们把 nextValue 就存在 invoker.value 中
          if (Array.isArray(invoker.value)) {
            // 如果是一个数组，说明一个事件绑定了多个方法，依次循环执行
            invoker.value.forEach((fn: any) => fn(e))
          } else {
            invoker.value(e)
          }
        }
        // 将事件存在 .value 中
        invoker.value = nextValue
        // .attached 储存事件处理函数被绑定的时间
        invoker.attached = performance.now()
        // 将该伪造事件添加为事件
        el.addEventListener(name, invoker)
      } else {
        // 如果存在，则更新 .value 的值即可
        invoker.value = nextValue
      }
    } else {
      // 如果不存在 nextValue ，说明已经不存在该方法了，则解除事件的绑定
      el.removeEventListener(name, invoker)
    }
  } else if (key === 'class') {
    // 如果是 class 则直接赋值
    el.className = nextValue || ''
  }
  // 判断是否需要设置 DOM Properties 的值
  else if (shouldSetAsProps(el, key, nextValue)) {
    // 获取该 DOM Properties 的类型
    const type = typeof el[key]
    // 如果是布尔类型，并且 value 是空字符串，则将值矫正为 true
    // 这是为了处理类似 button 的 disabled 等的属性的。这些属性无需设置值，只要存在该属性就需要为 true
    if (type === 'boolean' && nextValue === '') {
      el[key] = true
    } else {
      el[key] = false
    }
  } else {
    // 如果不存在对应的 DOM Properties 类型，则直接设置属性。
    el.setAttribute(key, nextValue)
  }
}

// 是否为 DOM Properties 设置值
function shouldSetAsProps(el: any, key: any, value: any) {
  // 因为 form 属性是只读，虽然 DOM Properties 上存在，但是还是只能通过 setAttribute 设置，所以需要特殊处理
  if (key === 'form' && el.tagName === 'INPUT') return false
  return key in el
}

const renderer = createRenderer({
  createElement,
  setElementText,
  insert,
  patchProps,
  unmount,
  createText,
  setText,
})

// 将设置好自定义方法配置的 renderer 实例封装，以提供给用户使用。
export function createApp(el: any, container: any) {
  renderer.render(el, container)
}

//因为 runtime-dom 是 runtime-core 的外层
//所以 runtime-core 在这里导出就好，打包的入口改为导出该 index.ts
export * from '../runtime-core'
