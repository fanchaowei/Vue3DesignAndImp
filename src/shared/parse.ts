export const namedCharacterReferences: any = {
  gt: '>',
  'gt;': '>',
  lt: '<',
  'lt;': '<',
  'ltcc;': 'âª¦'
}
