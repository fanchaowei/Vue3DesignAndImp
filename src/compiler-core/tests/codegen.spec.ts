import { generate } from '../src/codegen'
import { parse } from '../src/parse'
import { transforms } from '../src/transform'

describe('codegen', () => {
  test('happy path', () => {
    const ast: any = parse(`<div><p>Vue</p><p>Template</p></div>`)
    transforms(ast)
    const code = generate(ast.jsNode)
    expect(code).toMatchSnapshot()
  })
})
