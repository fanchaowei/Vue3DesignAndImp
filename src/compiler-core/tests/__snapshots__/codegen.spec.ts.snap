// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen happy path 1`] = `
"function render() {
  return h ('div',[h ('p','Vue'),h ('p','Template')])
}"
`;
